### ISSUE TITLE
UI corruptions caused by mpv

### ISSUE THREAD

**Issue by rien333** [xgi/castero#67]
Some files cause mpv to paint (unhelpful) warnings, and maybe also errors, over the castero interface (see screenshots). I have most warnings disabled in my mpv config, if that matters.

### Podcast 1

1. Add the following podcast feed to castero: https://anchor.fm/s/840a38c/podcast/rss
2.  Play the first episode:
![2019-11-17-1573948111](https://user-images.githubusercontent.com/10417027/69000689-064cb200-08d4-11ea-8d98-60d914dfd53b.jpg)

### Podcast 2

I found a podcast where this behavior is even more severe (but it has glitches all over, e.g. it displays some characters weirdly in the metadata field, so it might be not something fixable from castero). 

1. Add the following podcast feed: http://feeds.feedburner.com/CoolGamesInc?format=xml
2. Play the first episode
3. Scroll around a bit, and watch the interface glitch out
![2019-11-17-1573948266](https://user-images.githubusercontent.com/10417027/69000718-5b88c380-08d4-11ea-9213-5746cf1fb82b.jpg)

## System info

I'm running mpv and castero from git master, in kitty terminal.

**Comment by xgi** [comment]
Sorry for the late response. I was able to reproduce this, but I'm still working on a solution.

My current understanding is that these warnings are not from mpv itself (which should be and I believe are suppressed) but rather from ffmpeg. The line like `[mp3 @ 0x????????] Estimating duration...` is an ffmpeg warning, for example. I imagine there's a way to specify ffmpeg options (like loglevel) from mpv, but I'm still looking into that.

**Comment by rien333** [comment]
@xgi Tried to look into this. I tried disabling ffmpeg  output with  `--msg-level` on the commandline, and that works fine: 

```console
$ mpv https://chtbl.com/track/524GE/traffic.megaphone.fm/VMP8588299512.mp3
 [ffmpeg/demuxer]  mp3: Estimating duration from bitrate, this may be inaccurate
 (+) Video --vid=1 [P] 'image' (mjpeg 500x500)
...
$ mpv --msg-level=ffmpeg=no https://chtbl.com/track/524GE/traffic.megaphone.fm/VMP8588299512.mp3
 (+) Video --vid=1 [P] 'image' (mjpeg 500x500)
 (+) Audio --aid=1 (mp3 2ch 44100Hz)
```

However, when I try to set the msg-level via libmpv's mpv constructor (see [the python-mpv  README.md](https://github.com/jaseg/python-mpv/tree/3d6155871447ef5c5e4d54c39eb8ff573a3b49e9#pyqt-embedding)) in `castero/mpvplayer.py`, it still outputs ffmpeg warnings/errors:

```python
class MPVPlayer(Player):

...

def _create_player(self) -> None:
    """Creates the player object while making sure it is a valid file.

    Overrides method from Player; see documentation in that class.
    """
    self._player = self.mpv.MPV(log_handler=print, loglevel='fatal') # loglevel=no, error etc. behaves the same
```

Same when I set it directly (though I'm not sure if this code is correct, it doesn't error though, and is saved in `self._player['msg-level']`):

``` python
def _create_player(self) -> None:
    """Creates the player object while making sure it is a valid file.

    Overrides method from Player; see documentation in that class.
    """
    self._player = self.mpv.MPV()
    self._player.command("set", "msg-level", "ffmpeg=no")
```

Could this be a bug in `python-mpv`?

**Comment by rien333** [comment]
@jaseg any input?

**Comment by rien333** [comment]
> My current understanding is that these warnings are not from mpv itself (which should be and I believe are suppressed) but rather from ffmpeg. The line like [mp3 @ 0x????????] Estimating duration... is an ffmpeg warning, for example. 

Your belief about the warning being generated by ffmpeg (rather than mpv) is almost certainly correct: I think I've tried every option to disable mpv output, but the demuxer warning is still shown. Additionally, the warning message is colored. The `python-mpv` never would generated colored messages, except when explicitly told to do so.  

> I imagine there's a way to specify ffmpeg options (like loglevel) from mpv, but I'm still looking into that.

Unfortunately, that doesn't seem to be the case, looking at the mpv manual. There seem to be various ways to control ffmpeg filters, but no way to pass option like to it, like you would on the command line. 

That said, the issue might be due to code in castero. If I try to use `python-mpv` in a more minimal environment on one of the warning generating mp3 streams, I'm successfully able to silence the warning: 

```python
import mpv 

def log(loglevel, component, message):
    print('[{}] {}: {}'.format(loglevel, component, message))

msg_level = "demux=no:ad=no:ffmpeg=no:ao=no:ffmpeg/demuxer=no"
player = mpv.MPV(log_handler=log, msg_level=msg_level)

player.play('https://chtbl.com/track/524GE/traffic.megaphone.fm/VMP8588299512.mp3')
player.wait_for_playback()
```

(output)
```bash
# Without setting msg_level, the log would first display 
# "[warn] ffmpeg/demuxer: mp3: Estimating duration from bitrate, this may be inaccurate"
[info] cplayer:      Video --vid=1 [P] 'image' (mjpeg 500x500)
[info] cplayer:  (+) Audio --aid=1 (mp3 2ch 44100Hz)
[info] display-tags: File tags:
...
```

Strangely enough,  initialization mpv with `MPV(log_handler=log, msg_level=msg_level)` in the `_create_player` constructor of castero does silence the ffmpeg warning. The warning  only occurs within castero, but not more generally within `python-mpv` code.

**Comment by jaseg** [comment]
> @jaseg any input on how to hide ffmpeg warnings with python-mpv?

This is perfect timing. I might have built just what you're looking for for the python-mpv tests yesterday: https://github.com/jaseg/python-mpv/blob/master/mpv-test.py#L26

This redirects the program's `stdin` and `stdout` file descriptors as used by libc into `/dev/null` while preserving python's `sys.stdin` and `sys.stdout`. This way python program can still print() as usual but libmpv and any other C dependencies loaded into the process are effectively muted.

According to python doc, this should work with your curses code if you do this redirection before calling [curses.setupterm](https://docs.python.org/3.7/library/curses.html#curses.setupterm). You can just copy over the relevant parts of my code there, as you won't need the context manager.

**Comment by jaseg** [comment]
> Strangely enough, initialization mpv with `MPV(log_handler=log, msg_level=msg_level)` in the `_create_player` constructor of castero does not silence the ffmpeg warning. The warning only occurs within castero, but not more generally within `python-mpv` code.

My educated guess is that this is because that warning message is due to the way you're feeding data into libmpv. When calling mpv from the command line with a file or URL argument, it knows the file size of the played resource and it can seek through the file. Maybe for some reason it can't (or doesn't want to) do that when you're using libmpv through python-mpv, and the mp3 decoder then can't accurately determine the length of the VBR file.

**Comment by rien333** [comment]
Great suggestions, thank you so much. I'll see if they help fixing this bug. 

> My educated guess is that this is because that warning message is due to the way you're feeding data into libmpv.

This also entered my mind, good guess probably. I don't know a lot about how castero deals with this yet. 

btw, after fooling around with python-mpv in projects like this: great and handy library! Very easy to use and nice job on making mpv "pythonic".

**Comment by rien333** [comment]
@xgi Looks like just adopting the stderr-redirecting part from the  python-mpv test  linked by jaseg fixes this issue. If I try to mess with stdout as well I run into trouble with curses (although castero doesn't call `curses.setupterm`, like jaseg mentioned). I'm not completely clear about the details (like the appropriate location for the redirection  code), so I probably can't fix the issue, but it should theoretically be pretty easy.

Sidenote: @jaseg's test implies that the fix would be linux specific, but I'm not sure. BSDs, like macOS, share a lot of the same file handling and redirection concepts (macOS uses libc.dylib, instead of libc.so, though). Not sure about Windows, however.

**Comment by jaseg** [comment]
The redirection should happen at program startup, before curses is initialized. The setupterm call I think should happen at the beginning of curses initialization.

The fix can probably be ported to windows, but I'm not sure what you'd have to change. On MacOS and other posix-like environments it should just work I think.

**Comment by xgi** [comment]
I'm a bit busy at the moment but I'll plan on making these changes on the weekend. Thanks a ton to @rien333 for the investigation, and to @jaseg for the updates & support.

**Comment by xgi** [comment]
Fixed with #76. For now stderr just redirects to a temporary file, although it might be worth adding some real logging in the future. Thanks!
